### PR TITLE
Assignment7+8 linq: Omar Facundo and Roberto Garcia

### PR DESCRIPTION
**Pull request by omar5o9**
@omar5o9 and @robertogarcia11

### CODE REVIEW

**Comment by Matthew-Powers** [comment]
I will review this

File: Assignment.Interfaces/IAddress.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
Comment (Matthew-Powers) [inline Assignment.Interfaces/IAddress.cs:3]: Can remove these using statements

File: Assignment.Tests/Assignment.Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    <PackageReference Include="coverlet.collector" Version="6.0.0">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+  </ItemGroup>
+
+	<ItemGroup>
+		<ProjectReference Include="..\Assignment\Assignment.csproj" />
+
+
```
Comment (Matthew-Powers) [inline Assignment.Tests/Assignment.Tests.csproj:28]: Nitpick, excess whitespace

File: Assignment.Tests/NodeTests.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿using GenericsHomework;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (Matthew-Powers) [inline Assignment.Tests/NodeTests.cs:6]: Can remove all using statements other than GenericsHomework

File: Assignment.Tests/NodeTests.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+{
+    [Fact]
+    public void Node_NextReferencePointToSecondNode_Success()
+    {
+        Node<int> node = new(1);
+        node.Append(3);
+        Assert.NotEqual(node.Data, node.Next.Data);
+    }
+
+
```
Comment (Matthew-Powers) [inline Assignment.Tests/NodeTests.cs:19]: Add more tests, remove excess whitespace if there is any

File: Assignment.Tests/SampleDataTests.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.IO;
```
Comment (Matthew-Powers) [inline Assignment.Tests/SampleDataTests.cs:6]: Can remove all of these using statements

File: Assignment/Assignment.csproj [hunk 6]
```diff
@@ -0,0 +1,14 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+
```
Comment (Matthew-Powers) [inline Assignment/Assignment.csproj:7]: Nitpick: Remove whitespace

File: Assignment/Node.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (Matthew-Powers) [inline Assignment/Node.cs:4]: Can remove using statements aside from System.Collections

File: Assignment/Node.cs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        do
+        {
+            yield return current;
+            current = current.Next;
+            count++;
+        } while (current != this && count < maximum);
+        
+    }
+
+
```
Comment (Matthew-Powers) [inline Assignment/Node.cs:102]: Nitpick: can remove this excess whitespace

File: Assignment/Node.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public void Append(T value)
+    {
+        if(Exists(value))
+        {
+            throw new InvalidOperationException(nameof(value));
+        }   
+        Node<T> newNode = new(value);
+        Node<T> cur = this;
```
Comment (Matthew-Powers) [inline Assignment/Node.cs:27]: I know this is an old assignment, but would be good to rename cur to currentNode
```suggestion
        Node<T> currentNode = this;
```

File: Assignment/People.csv
Comment (Matthew-Powers) [inline Assignment/People.csv:1]: Change the properties on the file to Copy if newer

Right click on your People.csv file, and click on Properties at the bottom. Then Under Copy to Output Directory, choose Copy if newer

File: Assignment/Person.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Linq;
+using System.Collections.Generic;
```
Comment (Matthew-Powers) [inline Assignment/Person.cs:2]: Can remove these using statements

File: Assignment/SampleData.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
```
Comment (Matthew-Powers) [inline Assignment/SampleData.cs:4]: Can remove these using statements

File: Directory.Build.Props [hunk 12]
```diff
@@ -1,9 +1,10 @@
...
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
-    <EnableNETAnalyzers>true</EnableNETAnalyzers>
-    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
-  </PropertyGroup>
-</Project>
\ No newline at end of file
+<Project>
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+	<LangVersion>12.0</LangVersion>
```
Comment (Matthew-Powers) [inline Directory.Build.Props:4]: Can you adjust this to be lined up with the other properties?

File: Assignment.Tests/SampleDataTests.cs [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    public SampleDataTests()
+    {
+        SampleData = new SampleData();
+    }
+
+
+    [Fact]
+    public void GetAggregateListOfStatesGivenPeopleCollection_EmptyList_ReturnsSuccessful()
+    {
+        List<IPerson> people = new List<IPerson>();
```
Comment (Matthew-Powers) [inline Assignment.Tests/SampleDataTests.cs:22]: This can just be new()
```suggestion
        List<IPerson> people = new();

```

**Review by Matthew-Powers — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
        Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌
--Right click on your People.csv file, and click on Properties at the bottom. Then Under Copy to Output Directory, choose Copy if newer
        Using LINQ, skip the first row in the People.csv. ✔
        Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔

    Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
        Use ISampleData.CsvRows for your data source. ✔
        Don't forget the list should be unique. ✔
        Sort the list alphabetically. ❌✔
--Might be sorting alphabetically, but I don't know for certain
        Include a test that leverages a hardcoded list of addresses. ❌
--Missing test
        Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌
--Missing test

    Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
     Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source.✔
        Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
     Given the array, consider using string.Join to combine the list into a single string. ✔

    Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
        Use ISampleData.CsvRows as the source of the data. ✔
        Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
        Be sure that Person.Address is also populated. ✔
        Adding null validation to all the Person and Address properties is optional.
        Consider using ISampleData.CsvRows in your test to verify your results. ❌
--Missing test

    Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
        Use ISampleData.People for your data source. ✔

    Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔
      Use the people parameter from ISampleData.People property for your data source. ✔
        At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ❌
--Not using method
        Don't forget the list should be unique. ✔
        It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.

    Given the implementation of Node in Assignment5

    Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
    Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum. ✔

Extra Credit

    Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

Fundamentals

    Place all shared project properties into a Directory.Build.Props file.✔
    Place all shared project items into a Directory.Build.targets file.❌
--File does not exist
    Ensure nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    and enabled .NET analyzers for both projects ❌
--Not in any of the project files, but is in the props file. If the project files use the props file, ignore this
    For this assignment, consider using Assert.AreEqual<T>() (the generic version) ❌
--Uses .Equal, but this might not matter
    All of the above should be unit tested ❌
--Testing almost none of the code
    Choose simplicity over complexity ✔

Need to fix a few things, and tons of tests are missing

File: Assignment.Tests/Assignment.Tests.csproj (hunk 2 above)
Comment (robertogarcia11) [reply Assignment.Tests/Assignment.Tests.csproj:28]: removed, thanks!

File: Assignment.Tests/SampleDataTests.cs (hunk 13 above)
Comment (robertogarcia11) [reply Assignment.Tests/SampleDataTests.cs:22]: fixed thanks!

File: Assignment/Assignment.csproj (hunk 6 above)
Comment (robertogarcia11) [reply Assignment/Assignment.csproj:7]: done, thanks!

File: Assignment/Node.cs (hunk 9 above)
Comment (robertogarcia11) [reply Assignment/Node.cs:27]: changed the name, thank you!

File: Assignment/Node.cs (hunk 8 above)
Comment (robertogarcia11) [reply Assignment/Node.cs:102]: done!

File: Assignment/People.csv
Comment (robertogarcia11) [reply Assignment/People.csv:1]: thanks for looking out, totally forgot to change this after we restarted!

File: Assignment.Interfaces/IAddress.cs (hunk 1 above)
Comment (robertogarcia11) [reply Assignment.Interfaces/IAddress.cs:3]: done!

File: Assignment.Interfaces/IPerson.cs [hunk 14]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
[the remark this answers is not in the document]
Comment (robertogarcia11) [reply Assignment.Interfaces/IPerson.cs:3]: done!

File: Assignment.Tests/NodeTests.cs (hunk 3 above)
Comment (robertogarcia11) [reply Assignment.Tests/NodeTests.cs:6]: thank you! removed

File: Assignment.Tests/NodeTests.cs (hunk 4 above)
Comment (robertogarcia11) [reply Assignment.Tests/NodeTests.cs:19]: will add more tests and remove whitespace

File: Assignment.Tests/SampleDataTests.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+        string states = SampleData.GetAggregateListOfStatesGivenPeopleCollection(people);
+        Assert.Equal("", states);
+    }
+
```
[the remark this answers is not in the document]
Comment (robertogarcia11) [reply Assignment.Tests/SampleDataTests.cs:26]: currently adding more thank you!

File: Assignment/Node.cs (hunk 7 above)
Comment (robertogarcia11) [reply Assignment/Node.cs:4]: thank you!

File: Assignment/Person.cs (hunk 10 above)
Comment (robertogarcia11) [reply Assignment/Person.cs:2]: thanks removed!

File: Assignment/SampleData.cs (hunk 11 above)
Comment (robertogarcia11) [reply Assignment/SampleData.cs:4]: done!

File: Directory.Build.Props (hunk 12 above)
Comment (robertogarcia11) [reply Directory.Build.Props]: done thank you!

File: Assignment.Tests/SampleDataTests.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (omar5o9) [reply Assignment.Tests/SampleDataTests.cs:42]: Added more tests :)

File: Assignment/People.csv
Comment (omar5o9) [reply Assignment/People.csv:1]: Thanks for catching

File: Assignment/Node.cs [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        Node<T> current = this;
+        int count = 0;
+        do
+        {
+            yield return current;
+            current = current.Next;
+            count++;
+        } while (current != this && count < maximum);
+
```
Comment (BenjaminMichaelis) [inline Assignment/Node.cs:96]: Why not use linq?

File: Assignment/SampleData.cs [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    // 3.
+    public string GetAggregateSortedListOfStatesUsingCsvRows()
+    {
+        return string.Join(", ", GetUniqueSortedListOfStatesGivenCsvRows().ToArray());
+    }
+
+    // 4.
+    public IEnumerable<IPerson> People
+        => CsvRows.Select(line => line.Split(",")).OrderBy(line => (line[5], line[6], line[7]))
+        .Select(line => new Person(line[1], line[2], new Address(line[4], line[5], line[6], line[7]), line[3]));
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:20]: for specific columns in this file, a class or enum would be nice to give readable values instead of like '6' or '2' to different columns, so you could say like `[Column.State]` or `[Column.City]`, etc.

File: Assignment.Tests/SampleDataTests.cs [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+    
+
+    
+
+
+
+
+}
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/SampleDataTests.cs:58]:
```suggestion
    }
}
```
nit: whitespace

File: Assignment.Tests/Assignment.Tests.csproj [hunk 19]
```diff
@@ -0,0 +1,26 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/Assignment.Tests.csproj:6]: Why isn't this just in the `props` file?

File: Assignment/Node.cs [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+        do
+        {
+            yield return current;
+            current = current.Next;
+        } while (current != this);
+    }
+
+    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+    public IEnumerator<Node<T>> ChildItems(int maximum)
```
Comment (BenjaminMichaelis) [inline Assignment/Node.cs:86]: Why are there no tests around this?
